nixos/modules/services/cluster/kubernetes/dns.nix

```diff
@@ -27,232 +32,227 @@ in {
       type = types.str;
     };
 
-    kube-dns = mkOption {
-      description = "Docker image to seed for the kube-dns main container.";
-      type = types.attrs;
-      default = {
-        imageName = "k8s.gcr.io/k8s-dns-kube-dns-amd64";
-        imageDigest = "sha256:b99fc3eee2a9f052f7eb4cc00f15eb12fc405fa41019baa2d6b79847ae7284a8";
-        finalImageTag = version;
-        sha256 = "0x583znk9smqn0fix7ld8sm5jgaxhqhx3fq97b1wkqm7iwhvl3pj";
-      };
-    };
-
-    dnsmasq-nanny = mkOption {
-      description = "Docker image to seed for the kube-dns dnsmasq container.";
-      type = types.attrs;
-      default = {
-        imageName = "k8s.gcr.io/k8s-dns-dnsmasq-nanny-amd64";
-        imageDigest = "sha256:bbb2a290a568125b3b996028958eb773f33b5b87a6b37bf38a28f8b62dddb3c8";
-        finalImageTag = version;
-        sha256 = "1fihml7s2mfwgac51cbqpylkwbivc8nyhgi4vb820s83zvl8a6y1";
-      };
+    replicas = mkOption {
+      description = "Number of DNS pod replicas to deploy in the cluster.";
+      default = 2;
```

Comment:
What happens on a one-node cluster?

Reply:
Upstream doesn't touch this: https://github.com/kubernetes/kubernetes/blob/v1.12.2/cluster/addons/dns/coredns/coredns.yaml.sed#L85-L88 instead it states it's handled at runtime by horizontal auto-scaling.

Reply:
@steveeJ A one-node cluster will have 2 replicas deployed (by default), but both on the same node. I understand that scaling can be handled by horizontal auto-scalers, but **1)** we don't deploy the horizontal auto-scaler with this module (as of now), and **2)** I prefer multiple replicas not for scaling reasons, but for redundancy reasons. In our production cluster, we cannot tolerate DNS-service downtime due to the failure of a single container (or node). I can accept setting replicas to 1 by default, but in that case we really should set _maxUnavailable: 0_  to ensure that at least 1 DNS pod is always up during rolling-update of the DNS-deployment.

But that would still not ensure redundancy in case of node-failure. In our production cluster we have 12 nodes now, and we want to be able to guarantee zero downtime of the DNS-service, even in the case we have multiple nodes failing. We therefore currently set `replicas = 4`.

Reply:
Thanks for explaining. My remarks have been resolved, not to say invalidated ;-)